pilot/model/controller_view.go

```diff
@@ -0,0 +1,580 @@
+	// ipv4/v6 address, example 10.1.1.3, the value associated with
+	// this internal label is a normalized IPv6 address encoded in hex
+	labelServiceVIP = labelServicePrefix + "vip"
+
+	// Prefix for labels used internally by ControllerView for pertaining to
+	// to Service Instance properties
+	labelInstancePrefix = labelPilotPrefix + "ServiceInstance."
+
+	// Label used internally by ControllerView to fetch instances
+	// by instance IP. While ServiceDiscovery function parms may contain
+	// ipv4/v6 address, example 10.1.1.3, the value associated with
+	// this internal label is a normalized IPv6 address encoded in hex
+	labelInstanceIP = labelInstancePrefix + "ip"
+
+	// Label used internally by ControllerView to fetch instances
+	// by Port. While ServiceDiscovery function parms may contain
+	// alternate numeric formats for port numbers, ex: 80 or 0080
+	// etc, the value associated with this internal label is a
+	// normalized 16 bit int encoded in hex
+	labelInstancePort = labelInstancePrefix + "port"
+
+	// Label used internally by ControllerView to fetch instances
+	// by the named port.
+	labelInstanceNamedPort = labelInstancePrefix + "namedPort"
+)
+
+// ControllerView implements common storage of mesh entities like Service
+// and ServiceInstance objects and is intended to be used by individual
+// platform registry Controllers. Controller implementations call Reconcile()
+// to update the Controller's view with the most up-to-date list of mesh
+// entities available natively for that platform registry.
+//
+// TODO adopt ControllerView into all platform registry implementations as part
+// of https://github.com/istio/istio/issues/1223.
+//
+// Although objects from each Controller are distinct, they implicitly follow the
+// following contraints:
+//
+// - Service objects: are unique across the mesh for the combination of their
+// Hostname and Address
+// - ServiceInstance objects: are unique across the mesh for the combination
+// of their Service.Hostname and Endpoint.Address and Endpoint.Port
+//
+// The ControllerView will never modify model.* objects. However to accommodate
+// multiple data representations of various properties of model.* objects, the
+// ControllerView normalizes the following properties to allow for uniform
+// query operations across the mesh:
+//
+// - Address: Fetches by Service and ServiceInstance addresses are
+// always normalized to their 16 byte IPv6 representation.
+// - Port: Given that 0080 and 80 are equivalent, port numbers are always normalized
+// by their numeric 16 bit value.
+type ControllerView struct {
+
+	// serviceLabels holds the inverted map of label name/values to the keys in
+	// services for O(N) access involving labels versus O(N^2) access via services
+	serviceLabels viewLabels
+
+	// services is a map maintained by the Controller with a composite key of
+	// Service.Hostname and Service.Address. The key is expected to be mesh unique
```

Comment:
What does it mean ? Can you add an example of the key ?

Reply:
And why is this needed?